curvefs/src/client/rpcclient/cli2_cient.cpp

```diff
@@ -0,0 +1,136 @@
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+
+/*
+ * Project: curve
+ * Created Date: Mon Sept 4 2021
+ * Author: lixiaocui
+ */
+
+#include "curvefs/src/client/rpcclient/cli2_client.h"
+
+namespace curvefs {
+namespace client {
+namespace rpcclient {
+
+void Cli2Closure::Run() {
+    bool success = false;
+    if (cntl.Failed()) {
+        LOG(WARNING) << "GetLeader failed from " << cntl.remote_side()
+                     << ", logicpool id = " << taskContext.poolID
+                     << ", copyset id = " << taskContext.copysetID
+                     << ", error = " << cntl.ErrorText();
+    } else {
+        LOG(INFO) << "GetLeader returned from " << cntl.remote_side()
+                  << ", logicpool id = " << taskContext.poolID
+                  << ", copyset id = " << taskContext.copysetID
+                  << ", leader = " << response.DebugString();
+        success = true;
+    }
+
+    excutor->NotifyRpcFinish(success);
+}
+
+#define RPCFunc [&](brpc::Channel * channel) -> int  // NOLINT
+
+bool Cli2ClientImpl::GetLeader(LogicPoolID poolID, CopysetID copysetID,
+                               const PeerInfoList &peerInfoList,
+                               int16_t currentLeaderIndex, PeerAddr *peerAddr,
+                               MetaserverID *metaserverID) {
+    int16_t index = -1;
+    peerAddr->Reset();
+
+    // std::unordered_set<std::string> peerAddrs;
+    bool getLeaderOK = false;
+    for_each(
+        peerInfoList.begin(), peerInfoList.end(),
+        [&](const CopysetPeerInfo<MetaserverID> &info) {
+            ++index;
+            // skip current index
+            if (index != currentLeaderIndex && !getLeaderOK) {
+                // make request
+                auto excutor = std::make_shared<TaskExcutor2>();
+                Cli2Closure *done = new Cli2Closure(
+                    Cli2TaskContext(
+                        poolID, copysetID,
+                        butil::endpoint2str(info.externalAddr.addr_).c_str()),
+                    excutor);
+                done->cntl.set_timeout_ms(opt_.rpcTimeoutMs);
+
+                auto task = RPCFunc {
+                    curvefs::metaserver::copyset::GetLeaderRequest2 request;
+                    request.set_poolid(poolID);
+                    request.set_copysetid(copysetID);
+
+                    curvefs::metaserver::copyset::CliService2_Stub stub(
+                        channel);
+                    stub.GetLeader(&done->cntl, &request, &done->response,
+                                   done);
+                };
+
+                // analyse response
+                if (true == excutor->DoRPCTaskAndWait(
```

Comment:
`DoRPCTaskAndWait` make get leader from different peers serially, so if first rpc is timedout,  then it send anthoer rpc to second peer, the whole get leader process will at least cost `opt_.rpcTimeoutMs`

Reply:
yes, it is.